doc/design/thin_pool.md

```diff
@@ -0,0 +1,123 @@
+
++   // ThinPoolConfig contains configurations for the thin-pool
++   ThinPoolConfig  ThinPoolConfig `json:"thinPoolConfig,omitempty"`
+}
+```
+
+
+- Following new fields will added to `DeviceClass` API
+    - **ThinPoolConfig** API contains information related to a thin pool.These configuration options are:
+        - **Name**: Name of the thin-pool
+        - **SizePercent**: Size of the thin pool to be created with respect to available free space in the volume group. It represents percentage value and not absolute size values. Size value should range between 10-90. It defaults to 75 if no value is provided.
+        - **OverprovisionRatio**: The factor by which additional storage can be provisioned compared to the available storage in the thin pool.
+
+- `LVMVolumeGroup` API changes:
+
+``` go=
+type LVMVolumeGroupSpec struct {
+    // DeviceSelector is a set of rules that should match for a device to be
+    // included in this VolumeGroup
+	// +optional
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+
+    // NodeSelector chooses nodes
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+
++   // ThinPoolConfig contains configurations for the thin-pool
++   ThinPoolConfig ThinPoolConfig `json:"thinPoolConfig,omitempty"`
+}
+```
+
+### VolumeGroup Manager
+- Volume Group Manager is responsible for creating the thin-pools after creating the volume group.
+- Command used for creating the thin-pool:
+    ```
+    lvcreate -L <Size>%FREE -c <Chunk size> -T <vg_name>/<thin-pool-name>
+    ```
+    where:
+    - Size is `LVMClusterSpec.Storage.DeviceClass.ThinPoolConfig.SizePercent`
+    - chunk size is 512KiB, which is the default.
+
+- VG manager will also update the `lvmd.yaml` file to map volume group and its thin-pool to the topolvm device class.
+- Sample `lvmd.yaml` config file
+``` yaml=
+device-classes:
+  - name: ssd-thin
+    volume-group: myvg1
+    spare-gb: 10
+    type: thin
+    thin-pool-config:
+        name: pool0
+        overprovision-ratio: 50.0
+```
+
+### Monitoring and Alerts
+- Available thin-pool size (both data and metadata) should be provided by topolvm as prometheus metrics.
+- Threshold limits for the thin-pool should be provide as static values in the prometheus yaml file.
+- If available size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.
+
+
+### Open questions
+- What should be the chunk size of the thin-pools?
+    - Use default size a 512 kiB for now.
+- Should LVMO notify changes of thin-pool if it happens to topolvm?
+- How should vg-manager handle expansion of thin-pool in near future?
+- After firing alerts should LVMO take any pre-cautionary steps to make thin-lvs occupying whole thin-pool without effecting snapshots
+- Do we want to leverage dm-eventd for auto-expansion of thinpool, how could Topolvm monitoring and dm-eventd be made to work in tandem
```

Comment:
- as we'll be using topolvm solely for monitoring, no need for dm-eventd for auto-expansion, this question can be dropped